File: java/src/main/java/exercises/FixedArray.java
package exercises;
 
/**
 * Exercise 1: Fixed-Size Array
 * =============================
 * Arrays have a FIXED capacity allocated upfront. Once full, you cannot add
 * more elements without allocating a new array. This is a fundamental
 * constraint that shapes every decision about when to use them.
 *
 * Java's built-in arrays are already fixed-size, but here you will wrap one
 * to practice the insert/delete/shift logic explicitly.
 *
 * Your task: implement all methods marked with TODO.
 */
public class FixedArray {
 
    private final Object[] data;
    private final int capacity;
    private int size;
 
    /**
     * Initialize a fixed-size array with the given capacity.
     * The underlying array is pre-allocated — this models contiguous
     * memory allocation where the space is reserved upfront.
     */
    public FixedArray(int capacity) {
        this.capacity = capacity;
        this.data = new Object[capacity];
        this.size = 0;
    }
 
    /**
     * Insert {@code value} at position {@code index}.
     *
     * - Shift all elements from {@code index} onward one position to the right.
     * - Place {@code value} at {@code index}.
     * - Increment size.
     *
     * @throws IndexOutOfBoundsException if index < 0 or index > current size
     * @throws IllegalStateException     if the array is already at capacity
     *
     * Think about: why is shifting O(n)? What does this mean for insert-heavy workloads?
     */
    public void insert(int index, Object value) {
        if (index < 0 || index > size) {
            throw new IndexOutOfBoundsException();
        }
        if (size == capacity) {
            throw new IllegalStateException("Array is at full capacity");
        }
 
        for (int i = size - 1; i >= index; i--) {
            data[i + 1] = data[i];
        }
 
        data[index] = value;
        size++;
    }
 
    /**
     * Return the value at {@code index}.
     *
     * @throws IndexOutOfBoundsException if index < 0 or index >= current size
     *
     * This is O(1) — the key advantage of arrays over linked lists.
     */
    public Object get(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException();
        }
        return data[index];
    }
 
    /**
     * Delete the element at {@code index}.
     *
     * - Shift all elements after {@code index} one position to the left.
     * - Null out the last occupied slot.
     * - Decrement size.
     *
     * @throws IndexOutOfBoundsException if index < 0 or index >= current size
     */
    public void delete(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException();
        }
 
        for (int i = index; i < size - 1; i++) {
            data[i] = data[i + 1];
        }
 
        data[size - 1] = null;
        size--;
    }
 
    /**
     * Return the index of the first occurrence of {@code value}.
     * Return -1 if not found.
     *
     * This is O(n) — there is no shortcut without sorting.
     */
    public int find(Object value) {
        for (int i = 0; i < size; i++) {
            if (value == null ? data[i] == null : value.equals(data[i])) {
                return i;
            }
        }
        return -1;
    }
 
    /** Return the number of occupied slots (not the capacity). */
    public int size() {
        return size;
    }
 
    /**
     * Rotate the occupied elements of the array to the right by {@code k} positions.
     *
     * Example: [1, 2, 3, 4, 5] rotated right by 2 → [4, 5, 1, 2, 3]
     *
     * Only rotate the occupied portion (indices 0 to size-1).
     * k may be larger than size — handle this with modulo arithmetic.
     *
     * Think about: what data structure property makes this an in-place operation?
     */
    public void rotateRight(int k) {
        if (size == 0) return;
 
        k = k % size;
        if (k == 0) return;
 
        reverse(0, size - 1);
        reverse(0, k - 1);
        reverse(k, size - 1);
    }
 
    private void reverse(int left, int right) {
        while (left < right) {
            Object temp = data[left];
            data[left] = data[right];
            data[right] = temp;
            left++;
            right--;
        }
    }
 
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("FixedArray(capacity=" + capacity + ", elements=[");
        for (int i = 0; i < size; i++) {
            sb.append(data[i]);
            if (i < size - 1) sb.append(", ");
        }
        sb.append("])");
        return sb.toString();
    }
 
    // ✅ Main method added ONLY for manual testing / demo
    public static void main(String[] args) {
        FixedArray arr = new FixedArray(5);
 
        arr.insert(0, 1);
        arr.insert(1, 2);
        arr.insert(2, 3);
        arr.insert(1, 99);
 
        System.out.println(arr);
 
        arr.delete(2);
        System.out.println("After delete: " + arr);
 
        arr.rotateRight(1);
        System.out.println("After rotateRight(1): " + arr);
 
        System.out.println("Index of 99: " + arr.find(99));
        System.out.println("Current size: " + arr.size());
    }
}
 
 

File: cpp/exercises/fixed_array.cpp
#include "fixed_array.h"
#include <sstream>

FixedArray::FixedArray(int capacity)
    : data_(new int[capacity]), capacity_(capacity), size_(0) {}

FixedArray::~FixedArray() {
    delete[] data_;
}

void FixedArray::insert(int index, int value) {
    // TODO: implement this
    // 1. If index < 0 or index > size_, throw std::out_of_range("index out of range")
    // 2. If size_ == capacity_, throw std::overflow_error("array is full")
    // 3. Shift elements at [index .. size_-1] one position right
    // 4. Set data_[index] = value; increment size_
    if (index < 0 || index > size_) {
        throw std::out_of_range("index out of range");
    }


    if (size_ == capacity_) {
        throw std::overflow_error("array is full");
    }

    for (int i = size_; i > index; --i) {
        data_[i] = data_[i - 1];
    }

    data_[index] = value;
    ++size_;
}

int FixedArray::get(int index) const {
    // TODO: implement this
    // 1. If index < 0 or index >= size_, throw std::out_of_range("index out of range")
    // 2. Return data_[index]
    if (index < 0 || index >= size_) {
        throw std::out_of_range("index out of range");
    }

    // 2. Return element
    return data_[index];
    return 0; // placeholder — remove when implemented
}

void FixedArray::deleteAt(int index) {
    // TODO: implement this
    // 1. If index < 0 or index >= size_, throw std::out_of_range("index out of range")
    // 2. Shift elements at [index+1 .. size_-1] one position left
    // 3. Decrement size_
    if (index < 0 || index >= size_) {
        throw std::out_of_range("index out of range");
    }

    // 2. Shift elements left
    for (int i = index; i < size_ - 1; ++i) {
        data_[i] = data_[i + 1];
    }

    // 3. Decrement size
    --size_;
}

int FixedArray::find(int value) const {
    // TODO: implement this
    // Linear scan: return the index of the first element equal to value, or -1
    for (int i = 0; i < size_; ++i) {
        if (data_[i] == value) {
            return i;
        }
    }

    return -1;
    return -1; // placeholder — remove when implemented
}

int FixedArray::size() const {
    return size_;
}

void reverseRange(int* arr, int start, int end) {
    while (start < end) {
        int temp = arr[start];
        arr[start] = arr[end];
        arr[end] = temp;
        ++start;
        --end;
    }
}

void FixedArray::rotateRight(int k) {
    // TODO: implement this
    // Rotate the occupied region [0 .. size_-1] right by k positions.
    // Hint: use the "three-reverse" trick —
    //   1. Normalise: k = k % size_ (guard against size_==0 first)
    //   2. Reverse the entire occupied region
    //   3. Reverse the first k elements
    //   4. Reverse elements [k .. size_-1]
    if (size_ == 0) {
        return;
    }

    // Normalize k
    k = k % size_;

    if (k == 0) {
        return;
    }

    // Three-reverse trick
    reverseRange(data_, 0, size_ - 1);
    reverseRange(data_, 0, k - 1);
    reverseRange(data_, k, size_ - 1);
}

std::string FixedArray::toString() const {
    std::ostringstream oss;
    oss << "[";
    for (int i = 0; i < size_; ++i) {
        if (i > 0) oss << ", ";
        oss << data_[i];
    }
    oss << "]";
    return oss.str();
}
